src/bin/kawari-world.rs

```diff
@@ -68,20 +74,368 @@ fn spawn_main_loop() -> (ServerHandle, JoinHandle<()>) {
+/// Spawn a new chat connection for an incoming client.
+fn spawn_chat_connection(connection: ChatConnection) {
+    let (send, recv) = channel(64);
+
+    let id = &connection.id.clone();
+    let ip = &connection.ip.clone();
+
+    let data = ClientChatData { recv, connection };
+
+    // Spawn a new client task
+    let (my_send, my_recv) = oneshot::channel();
+    let _kill = tokio::spawn(start_chat_connection(my_recv, data));
+
+    // Send client information to said task
+    let handle = ClientHandle {
+        id: *id,
+        ip: *ip,
+        channel: send,
+        actor_id: 0,
+    };
+    let _ = my_send.send(handle);
+}
+
+/// The task that kickstarts the client chat connection loop.
+async fn start_chat_connection(my_handle: oneshot::Receiver<ClientHandle>, data: ClientChatData) {
+    // Recieve client information from global
+    let my_handle = match my_handle.await {
+        Ok(my_handle) => my_handle,
+        Err(_) => return,
+    };
+
+    let connection = data.connection;
+    let recv = data.recv;
+
+    // communication channel between client_loop and client_server_loop
+    let (internal_send, internal_recv) = unbounded_channel();
+
+    let _ = join!(
+        tokio::spawn(client_chat_loop(connection, internal_recv, my_handle)),
+        // TODO: It should be okay to reuse the zone connection's client_server_loop for ToServer & FromServer stuff since it gives us our own comms channel, right?
+        tokio::spawn(client_server_loop(recv, internal_send))
+    );
+}
+
+/// The client's chat connection loop, which allows them to communicate with other players in channels that span across the entire world.
+async fn client_chat_loop(
+    mut connection: ChatConnection,
+    mut internal_recv: UnboundedReceiver<FromServer>,
+    client_handle: ClientHandle,
+) {
+    let mut buf = vec![0; RECEIVE_BUFFER_SIZE];
+    loop {
+        tokio::select! {
+            biased; // client data should always be prioritized
+
+            n = connection.socket.read(&mut buf) => {
+                match n {
+                    Ok(n) => {
+                        // if the last response was over >5 seconds, the client is probably gone
+                        if n == 0 {
+                            let now = Instant::now();
+                            if now.duration_since(connection.last_keep_alive) > Duration::from_secs(5) {
+                                tracing::info!("Connection {:#?} was killed because of timeout", client_handle.id);
+                                break;
+                            }
+                        } else {
+                            connection.last_keep_alive = Instant::now();
+                            let (segments, _) = connection.parse_packet(&buf);
+                            for segment in segments {
+                                match &segment.data {
+                                    SegmentData::None() => {}
+                                    SegmentData::Setup {.. } => {
```

Comment:
```suggestion
                                    SegmentData::Setup { .. } => {
```

(Because `tokio::select!` is a macro, the formatter doesn't affect anything here :smile:

Reply:
Done!